File: backend/api/schemas/retrieval_test.py
"""Retrieval Testing endpoint schemas (ADR-0021 — V4 Retrieval Testing tab §5.5.4).

`POST /kb/{kb_id}/retrieval-test` runs a pure retrieval pass (no CRAG, no LLM
synthesis) so the admin can compare Vector / Full-Text / Hybrid modes, tune
Top-K / score threshold, and toggle rerank — surfacing ranked chunks + scores.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class RetrievalTestRequest(BaseModel):
    query: str = Field(..., min_length=1, max_length=2000)
    mode: Literal["hybrid", "vector", "fulltext"] = "hybrid"
    top_k: int = Field(5, ge=1, le=50)
    rerank: bool = True
    # Similarity threshold for vector / hybrid scores (0–1). Ignored for fulltext —
    # BM25 scores are unbounded above and have no 0–1 meaning (the frontend disables
    # this control when Full-Text mode is selected).
    score_threshold: float = Field(0.0, ge=0.0, le=1.0)


class RetrievalTestChunk(BaseModel):
    rank: int
    chunk_id: str
    doc_id: str
    doc_title: str
    chunk_title: str
    chunk_index: int
    section_path: list[str]
    score: float
    chunk_text_preview: str


class RetrievalTestResult(BaseModel):
    kb_id: str
    query: str
    mode: str
    reranked: bool
    reranker: str  # "cohere-v4.0-pro" when reranked, else "none"
    embed_latency_ms: int
    search_latency_ms: int
    rerank_latency_ms: int
    total_latency_ms: int
    total_hits: int  # before the score_threshold filter
    chunks: list[RetrievalTestChunk]

File: frontend/lib/api/retrieval-test.ts
/**
 * Typed Retrieval Testing API (ADR-0021 — V4 Retrieval Testing tab §5.5.4).
 *
 * POST /kb/{kb_id}/retrieval-test — pure retrieval (no CRAG, no LLM synthesis):
 * compare Vector / Full-Text / Hybrid modes, tune Top-K / score threshold,
 * toggle rerank → ranked chunks + relevance scores + per-stage timings.
 */

import { ApiClient } from '../api-client';

const client = new ApiClient();

export type RetrievalMode = 'hybrid' | 'vector' | 'fulltext';

export interface RetrievalTestRequest {
  query: string;
  mode?: RetrievalMode;
  top_k?: number;
  rerank?: boolean;
  /** 0–1 similarity threshold for vector/hybrid scores; ignored for fulltext (BM25 unbounded). */
  score_threshold?: number;
}

export interface RetrievalTestChunk {
  rank: number;
  chunk_id: string;
  doc_id: string;
  doc_title: string;
  chunk_title: string;
  chunk_index: number;
  section_path: string[];
  score: number;
  chunk_text_preview: string;
}

export interface RetrievalTestResult {
  kb_id: string;
  query: string;
  mode: string;
  reranked: boolean;
  reranker: string; // "cohere-v4.0-pro" when reranked, else "none"
  embed_latency_ms: number;
  search_latency_ms: number;
  rerank_latency_ms: number;
  total_latency_ms: number;
  total_hits: number; // before the score_threshold filter
  chunks: RetrievalTestChunk[];
}

export const retrievalTestApi = {
  run: (kbId: string, body: RetrievalTestRequest): Promise<RetrievalTestResult> =>
    client.post<RetrievalTestResult>(
      `/kb/${encodeURIComponent(kbId)}/retrieval-test`,
      body,
    ),
};
